[1.21.3] Fix unbound tags issue in registry snapshot application

dhyces opened the pull request:
Pretty much part 2 of #1651, due to missed registry freeze issue that didn't crop up in the added test. Hopefully the only other location where registry freeze needs to be messed with to allow named holder sets in static registry objects.

dhyces committed:
Fix registry snapshot application crash for tags

dhyces committed:
Update GameData.java

Shadows-of-Fire commented:
Can you try adding a test with a statically declared ArmorMaterial that uses a custom tag?  Trying to see if we can force the issue to be tested.

dhyces commented:
Tried a static variable, still succeeded just fine in the test after removing the fix. I'll look into whether there's some way to replicate this later.

dhyces commented:
> Can you try adding a test with a statically declared ArmorMaterial that uses a custom tag? Trying to see if we can force the issue to be tested.

Confirmed that this only seems to occur if a registry-related exception is thrown, causing the true error to be suppressed by the tag binding exception in freeze.

dhyces committed:
Alternative solution

I believe this solution to be more appropriate, as it removes effectively invalid named holder sets from the registry, rather than binding them

dhyces committed:
Spotless

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/registries/GameData.java:
```diff
@@ -66,7 +66,7 @@ public static void unfreezeData() {
         BuiltInRegistries.REGISTRY.stream().filter(r -> r instanceof MappedRegistry).forEach(r -> {
-            // HolderSet.Named may be used for registry objects, vanilla binds these tags so freeze won't throw for unbound tags
+            // HolderSet.Named may be used for registry objects, vanilla binds these tags so freeze doesn't throw for unbound tags
             ((MappedRegistry<?>) r).bindAllTagsToEmpty();
```
With the patch in `MappedRegistry` above, do we still need the `bindAllTagsToEmpty` call here?  Are they trying to achieve the same purpose?

dhyces replied:
Different purposes. Binding all tags to empty here allows tags to be used and they'll be properly rebound on world load. The clear call is used for when a registry error occurs, where it just removes all unbound holder sets, effectively reducing them to the same result as you would get from `HolderSet.emptyNamed`.

Shadows-of-Fire replied:
Makes sense, are we certain there are no other problematic call sites where we call `freezeRegistry` that might need special treatment?

dhyces replied:
I ran a search after this last error and didn't see any other problematic call sites, but I'll double check when I get back

dhyces replied:
Everything looks good, I'll resolve this and re-request review

dhyces committed:
Merge branch '1.21.x' into fix/registry-snapshot-tags

Shadows-of-Fire reviewed (approved):
Hopefully this doesn't regress again in the future, unfortunate that the trigger is too difficult to exercise in testing.
